adapters/sovrn/sovrn.go

```diff
@@ -0,0 +1,274 @@
+package sovrn
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbs"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+type SovrnAdapter struct {
+	http *adapters.HTTPAdapter
+	URI  string
+}
+
+// Name - export adapter name */
+func (s *SovrnAdapter) Name() string {
+	return "sovrn"
+}
+
+// FamilyName used for cookies and such
+func (s *SovrnAdapter) FamilyName() string {
+	return "sovrn"
+}
+
+type sovrnParams struct {
+	TagID int `json:"tagid"`
+}
+
+func (s *SovrnAdapter) SkipNoCookies() bool {
+	return false
+}
+
+// Call send bid requests to sovrn and receive responses
+func (s *SovrnAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *pbs.PBSBidder) (pbs.PBSBidSlice, error) {
+	supportedMediaTypes := []pbs.MediaType{pbs.MEDIA_TYPE_BANNER}
+	sReq, err := adapters.MakeOpenRTBGeneric(req, bidder, s.FamilyName(), supportedMediaTypes, true)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sovrnReq := openrtb.BidRequest{
+		ID:   sReq.ID,
+		Imp:  sReq.Imp,
+		Site: sReq.Site,
+	}
+
+	// add tag ids to impressions
+	for i, unit := range bidder.AdUnits {
+		var params sovrnParams
+		err = json.Unmarshal(unit.Params, &params)
+		if err != nil {
+			return nil, err
+		}
+		sovrnReq.Imp[i].Secure = sReq.Imp[i].Secure
+		sovrnReq.Imp[i].TagID = strconv.Itoa(params.TagID)
+		sovrnReq.Imp[i].Banner.Format = nil // todo: need to do this?
+	}
+
+	reqJSON, err := json.Marshal(sovrnReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug := &pbs.BidderDebug{
+		RequestURI: s.URI,
+	}
+
+	if req.IsDebug {
+		debug.RequestBody = string(reqJSON)
+		bidder.Debug = append(bidder.Debug, debug)
+	}
+
+	httpReq, _ := http.NewRequest("POST", s.URI, bytes.NewReader(reqJSON))
+	httpReq.Header.Set("Content-Type", "application/json")
+	httpReq.Header.Set("User-Agent", sReq.Device.UA)
+	httpReq.Header.Set("X-Forwarded-For", sReq.Device.IP)
+	httpReq.Header.Set("Accept-Language", sReq.Device.Language)
+	httpReq.Header.Set("DNT", strconv.Itoa(int(sReq.Device.DNT)))
+
+	userID := strings.TrimSpace(sReq.User.BuyerUID)
+	if len(userID) > 0 {
+		httpReq.AddCookie(&http.Cookie{Name: "ljt_reader", Value: userID})
+	}
+	sResp, err := ctxhttp.Do(ctx, s.http.Client, httpReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug.StatusCode = sResp.StatusCode
+
+	if sResp.StatusCode == 204 {
+		return nil, nil
+	}
+
+	defer sResp.Body.Close()
```

Comment:
Correct me if I'm wrong, but... I think this should happen right after the `err != nil` check after `ctxhttp.Do`?

The Go docs don't mention any special case about 204s... and if you get one, this code will return without ever closing the body.

Reply:
I think you're right. I actually took this from the appnexus adapter, so maybe that needs to be updated there as well ;)

Reply:
thanks! Yes, it looks like it should be. I've been totally ignoring the legacy adapter code to focus on openrtb, so... if something looks like it's wrong in the APN `Call()` function, there's a good chance that it is.